src/algebra/semiconj.lean

```diff
@@ -0,0 +1,228 @@
+
+In the names of these operations, we treat `a` as the “left” argument,
+and both `x` and `y` as “right” arguments. This way most names in this
+file agree with the names of the corresponding lemmas for `commute a b
+= semiconj_by a b b`. As a side effect, some lemmas have only `_right`
+version.
+
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(h.pow_right 5).eq]`
+rather than just `rw [h.pow_right 5]`.
+-/
+
+universes u v
+
+open_locale smul
+
+/-- `x` is semiconjugate to `y` by `a`, if `a * x = y * a`. -/
+def semiconj_by {M : Type u} [has_mul M] :
+  M → M → M → Prop :=
+λ a x y, a * x = y * a
+
+namespace semiconj_by
+
+section semigroup
+
+variables {S : Type u}
+
+protected lemma eq [has_mul S] {a x y : S} (h : semiconj_by a x y) :
+  a * x = y * a := h
+
+variables [semigroup S] {a b x y z x' y' : S}
+
+@[simp] lemma mul_right (h : semiconj_by a x y) (h' : semiconj_by a x' y') :
+  semiconj_by a (x * x') (y * y') :=
+by unfold semiconj_by; assoc_rw [h.eq, h'.eq]
+
+lemma mul_left (ha : semiconj_by a y z) (hb : semiconj_by b x y) :
+  semiconj_by (a * b) x z :=
+by unfold semiconj_by; assoc_rw [hb.eq, ha.eq, mul_assoc]
+
+end semigroup
+
+section monoid
+
+variables {M : Type u} [monoid M]
+
+@[simp] lemma one_right (a : M) : semiconj_by a 1 1 := by rw [semiconj_by, mul_one, one_mul]
+
+@[simp] lemma one_left (x : M) : semiconj_by 1 x x := eq.symm $ one_right x
+
+lemma units_inv_right {a : M} {x y : units M} (h : semiconj_by a x y) :
+  semiconj_by a ↑x⁻¹ ↑y⁻¹ :=
+calc a * ↑x⁻¹ = ↑y⁻¹ * (y * a) * ↑x⁻¹ : by rw [units.inv_mul_cancel_left]
+          ... = ↑y⁻¹ * a              : by rw [← h.eq, mul_assoc, units.mul_inv_cancel_right]
+
+@[simp] lemma units_inv_right_iff {a : M} {x y : units M} :
+  semiconj_by a ↑x⁻¹ ↑y⁻¹ ↔ semiconj_by a x y :=
+⟨units_inv_right, units_inv_right⟩
+
+lemma units_inv_symm_left {a : units M} {x y : M} (h : semiconj_by ↑a x y) :
+  semiconj_by ↑a⁻¹ y x :=
+calc ↑a⁻¹ * y = ↑a⁻¹ * (y * a * ↑a⁻¹) : by rw [units.mul_inv_cancel_right]
+          ... = x * ↑a⁻¹              : by rw [← h.eq, ← mul_assoc, units.inv_mul_cancel_left]
+
+@[simp] lemma units_inv_symm_left_iff {a : units M} {x y : M} :
+  semiconj_by ↑a⁻¹ y x ↔ semiconj_by ↑a x y :=
+⟨units_inv_symm_left, units_inv_symm_left⟩
+
+@[simp] protected lemma map {N : Type v} [monoid N] (f : M →* N) {a x y : M} (h : semiconj_by a x y) :
+  semiconj_by (f a) (f x) (f y) :=
+by simpa only [semiconj_by, f.map_mul] using congr_arg f h
+
+theorem units_coe {a x y : units M} (h : semiconj_by a x y) :
+  semiconj_by (a : M) x y :=
+congr_arg units.val h
+
+theorem units_of_coe {a x y : units M} (h : semiconj_by (a : M) x y) :
+  semiconj_by a x y :=
+units.ext h
+
+@[simp] theorem units_coe_iff {a x y : units M} :
+  semiconj_by (a : M) x y ↔ semiconj_by a x y :=
+⟨units_of_coe, units_coe⟩
+
+@[simp] lemma pow_right {x y z : M} (h : semiconj_by x y z) :
```

Comment:
```suggestion
@[simp] lemma pow_right {a x y : M} (h : semiconj_by a x y) :
```
with corresponding changes in definition; also in `units_gpow_right`.

Reply:
Done.